src/audio/mux/mux_generic.c

```diff
@@ -316,49 +160,98 @@ static void demux_s32le(struct audio_stream *sink,
  */
 static void mux_s32le(struct audio_stream *sink,
 		      const struct audio_stream **sources, uint32_t frames,
-		      struct mux_stream_data *data)
+		      struct mux_stream_data *data, struct mux_look_up_entry *look_up)
 {
 	const struct audio_stream *source;
 	uint8_t i;
-	uint8_t j;
+	uint8_t source_ch;
 	uint8_t out_ch;
+	uint32_t offset;
+	int32_t *src;
 	int32_t *dst;
 	uint32_t dst_idx;
-	int64_t sample;
 
 	for (i = 0; i < frames; i++) {
 		for (out_ch = 0; out_ch < sink->channels; out_ch++) {
-			sample = 0;
-			for (j = 0; j < MUX_MAX_STREAMS; j++) {
-				source = sources[j];
-				if (!source)
-					continue;
-
-				sample += calc_sample_s32le(source,
-						i * source->channels,
-						data[j].mask[out_ch]);
-			}
+			source = sources[look_up[out_ch].stream_id];
+
+			if (!source || !look_up[out_ch].set)
+				continue;
+
+			source_ch = look_up[out_ch].in_ch;
+			offset = i * source->channels;
+			src = audio_stream_read_frag_s32(source,
+							 offset + source_ch);
 			dst_idx = i * sink->channels + out_ch;
 			dst = audio_stream_write_frag_s32(sink, dst_idx);
-			*dst = sat_int32(sample);
+			*dst = *src;
 		}
 	}
 }
 
-#endif /* CONFIG_FORMAT_S32LE */
+#endif /* CONFIG_FORMAT_S24LE CONFIG_FORMAT_S32LE */
 
 const struct comp_func_map mux_func_map[] = {
 #if CONFIG_FORMAT_S16LE
 	{ SOF_IPC_FRAME_S16_LE, &mux_s16le, &demux_s16le },
 #endif
 #if CONFIG_FORMAT_S24LE
-	{ SOF_IPC_FRAME_S24_4LE, &mux_s24le, &demux_s24le },
+	{ SOF_IPC_FRAME_S24_4LE, &mux_s32le, &demux_s32le },
 #endif
 #if CONFIG_FORMAT_S32LE
 	{ SOF_IPC_FRAME_S32_LE, &mux_s32le, &demux_s32le },
 #endif
 };
 
+void mux_prepare_look_up_table(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint8_t i;
+	uint8_t j;
+	uint8_t k;
+
+	/* Prepare look up table */
+	for (i = 0; i < cd->config.num_streams; i++) {
+		for (j = 0; j < PLATFORM_MAX_CHANNELS; j++) {
+			for (k = 0; k < PLATFORM_MAX_CHANNELS; k++) {
+				if (cd->config.streams[i].mask[j] & BIT(k)) {
+					/* MUX component has only one sink */
+					cd->proc_lookup[0][j].in_ch = k;
+					cd->proc_lookup[0][j].stream_id = i;
+					cd->proc_lookup[0][j].set = true;
+				}
+			}
+		}
+	}
+
+	
+	for (j = 0; j < PLATFORM_MAX_CHANNELS; j++) {
+		comp_info(dev, "cd->proc_lookup[0][%d]: in_ch: %d, stream_id: %d", j, cd->proc_lookup[0][j].in_ch, cd->proc_lookup[0][j].stream_id);
+	}
+}
+
+void demux_prepare_look_up_table(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint8_t i;
+	uint8_t j;
+	uint8_t k;
+
+	/* Prepare look up table */
+	for (i = 0; i < cd->config.num_streams; i++) {
+		for (j = 0; j < PLATFORM_MAX_CHANNELS; j++) {
+			for (k = 0; k < PLATFORM_MAX_CHANNELS; k++) {
+				if (cd->config.streams[i].mask[j] & BIT(k)) {
+					/* MUX component has only one sink */
+					cd->proc_lookup[i][j].in_ch = k;
+					cd->proc_lookup[i][j].stream_id = i;
+					cd->proc_lookup[i][j].set = true;
+				}
+			}
+		}
+	}
+}
+
```

Comment:
This is in the right direction, but we should go down to a lower level which is far simpler
```c
/* copy element */
struct copy_elem {
    void *dest;
    void *src;
    uint32_t dest_inc;  /* increment size for dest pointer when elem is next copied */
    uint32_t src_inc;  /* increment size for src pointer when elem is next copied */
    bool dest_wrap; /* dest needs reset after current frame block has been copied */
    bool src_wrap; /* src needs reset after current frame block has been copied */
}

int copy_s16(int num_frames)
{
   while (num_frames) {

        /* calculate mux number of frames that can be copied without overflow - set src_wrap and dest_wrap in copy_elems*/
        frames = calc_max_frames();

        /* copy directly since wrap has been pre-computed */
        for (i = 0; i < frames; i++) {
            for (j = 0; j < num_copy_elems; j++) {
                 *(int_16_t*)(elem[j].dest) = *(int16_t*)(elem[j].src);
                  elem[j].dest += elem[j].dest_inc;
                  elem[j.src += elem[j].src_inc;
            }

         /* check for wrap - loop copy_elems and reset buffers if src_wrap or dest_wrap is true */
        check_for_wrap();

         /* decrement frames */
        num_frames -= frames;
    }
}
```